daemon/image_delete.go

```diff
@@ -13,164 +12,306 @@ import (
 	"github.com/docker/docker/utils"
 )
 
-// FIXME: remove ImageDelete's dependency on Daemon, then move to graph/
-func (daemon *Daemon) ImageDelete(name string, force, noprune bool) ([]types.ImageDelete, error) {
-	list := []types.ImageDelete{}
-	if err := daemon.imgDeleteHelper(name, &list, true, force, noprune); err != nil {
+// ImageDelete deletes the image referenced by the given imageRef from this
+// daemon. The given imageRef can be an image ID, ID prefix, or a repository
+// reference (with an optional tag or digest, defaulting to the tag name
+// "latest"). There is differing behavior depending on whether the given
+// imageRef is a repository reference or not.
+//
+// If the given imageRef is a repository reference then that repository
+// reference will be removed. However, if there exists any containers which
+// were created using the same image reference then the repository reference
+// cannot be removed unless either there are other repository references to the
+// same image or force is true. Following removal of the repository reference,
+// force will be set to false (regardless of the given value), and the
+// referenced image itself will attempt to be deleted as described below but
+// quietly, meaning any image delete conflicts will cause the image to not be
+// deleted and the conflict will not be reported.
+//
+// There may be conflicts preventing deletion of an image and these conflicts
+// are divided into two categories grouped by their severity:
+//
+// Hard Conflict:
+// 	- a pull or build using the image.
+// 	- any descendent image.
+// 	- any running container using the image.
+//
+// Soft Conflict:
+// 	- any stopped container using the image.
+// 	- any repository tag or digest references to the image.
+//
+// The image cannot be removed if there are any hard conflicts and can be
+// removed if there are soft conflicts only if force is true.
+//
+// If prune is true, ancestor images will each attempt to be deleted quietly,
+// meaning any delete conflicts will cause the image to not be deleted and the
+// conflict will not be reported.
+//
+// FIXME: remove ImageDelete's dependency on Daemon, then move to the graph
```

Comment:
The previous FIXME had no explanation for why it hasn't been done or why it's difficult to do, so I added a short description of the problem. You're right, it would require some way for other systems to hold soft/hard references to images.